sky/shell/dart/dart_library_provider_files.cc

```diff
@@ -66,15 +66,40 @@ void DartLibraryProviderFiles::GetLibraryAsStream(
     blink::DataPipeConsumerCallback callback) {
   mojo::DataPipe pipe;
   callback.Run(pipe.consumer_handle.Pass());
-
-  base::FilePath source(name);
+  std::string path = GetFilePathForURL(name);
+  base::FilePath source(path);
   scoped_refptr<base::TaskRunner> runner =
       base::WorkerPool::GetTaskRunner(true);
   mojo::common::CopyFromFile(source, pipe.producer_handle.Pass(), 0,
                              runner.get(), base::Bind(&CopyComplete, source));
 }
 
-std::string DartLibraryProviderFiles::CanonicalizePackageURL(std::string url) {
+Dart_Handle DartLibraryProviderFiles::CanonicalizeURL(Dart_Handle library,
+                                                      Dart_Handle url) {
+  std::string string = blink::StdStringFromDart(url);
+  if (base::StartsWithASCII(string, "dart:", true))
+    return url;
+  if (base::StartsWithASCII(string, "package:", true))
+    return url;
+  if (base::StartsWithASCII(string, "file:", true))
+    return url;
+  base::FilePath base_path(blink::StdStringFromDart(Dart_LibraryUrl(library)));
+  base::FilePath resolved_path = base_path.DirName().Append(string);
+  base::FilePath normalized_path = SimplifyPath(resolved_path);
```

Comment:
We should probably be working in strings here rather than FilePaths.  These strings start with nutty things like "package:" that aren't sensible file paths.